### ISSUE TITLE
Change in order of columns 

### ISSUE THREAD

**Issue by gmbecker** [jeroenooms/jsonlite#64]
If an NA occurs in the first row of a data.frame, the original (obj) and fromJSON(toJSON(obj)) will have differently ordered columns

```
> thing
   a    b
1 NA    A
2  2 <NA>
> fromJSON(toJSON(thing))
     b  a
1    A NA
2 <NA>  2
```

**Comment by jeroenooms** [comment]
This is expected unfortunately. A json **object** is an **unordered** set of key-value pairs. Therefore, unlike in R, you must use refer to list elements or data frame columns by their name; the order is arbitrary.

**Comment by gmbecker** [comment]
I understand, but the data came (or might have) from R, so wouldn't it be
easy to have a top-level attribute that defined the column names and order?

~G

On Fri, Feb 13, 2015 at 3:03 PM, Jeroen Ooms <notifications@github.com>
wrote:

> This is expected unfortunately. A json *object* is an *unordered* set of
> key-value pairs. Therefore, unlike in R, you must use refer to list
> elements or data frame columns by their name; the order is arbitrary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jeroenooms/jsonlite/issues/64#issuecomment-74342805>.
>



-- 
Gabriel Becker, PhD
Computational Biologist
Bioinformatics and Computational Biology
Genentech, Inc.

**Comment by jeroenooms** [comment]
Nah, the `toJSON` and `fromJSON` functions are designed for interoperable data interchange, they should not contain R specific attributes or meta data. You are free to include this sort of stuff yourself, something like this:

```r
toJSON(
  list(data = mydata, order = names(mydata))
)
```

If you are looking for an out of the box solution to encode R objects in json that retains all meta data, have a look at the `serializeJSON` function.

**Comment by gmbecker** [comment]
Eh, my point was less about R-specific attributes and more about losing
semantic information about the object that is (on one side of the transfer)
relatively important.

I understand that associative arrays are unordered, but data.frames - and,
in general, tabular representations of data in many languages - aren't. My
concern is that if I am receiving data in chunks that come from a table,
the pieces I get won't be concordant as data.frames in R. And  the most
extreme case for chunked data is substantially worse:

> prettify(toJSON(data.frame(x=1:2, y = c(NA, NA)))
+ )
[
    {
        "x": 1
    },
    {
        "x": 2
    }
]

Ultimately, I think it has less to do with orderedness or R-specific
attributes *per se* and more to do with the representation of NA as absence.

I'm not claiming to have all the answers, I have struggled with these same
issues and I think jsonlite is a nice package. Just letting you know what
another set of eyes sees.

~G

On Sat, Feb 14, 2015 at 2:13 PM, Jeroen Ooms <notifications@github.com>
wrote:

> Closed #64 <https://github.com/jeroenooms/jsonlite/issues/64>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jeroenooms/jsonlite/issues/64#event-235900366>.
>



-- 
Gabriel Becker, PhD
Computational Biologist
Bioinformatics and Computational Biology
Genentech, Inc.

**Comment by jeroenooms** [comment]
Yes, I've thought a lot about this, there is no one size fits all solution. The defaults of `toJSON` are designed to optimize interoperability. If you are concerned about missing values getting lost, you can encode these explicitly in the data by setting the `na` argument:

```r
toJSON(data.frame(x=1:2, y = c(NA, NA)), na = "string")
# [{"x":1,"y":"NA"},{"x":2,"y":"NA"}] 

toJSON(data.frame(x=1:2, y = c(NA, NA)), na = "null")
# [{"x":1,"y":null},{"x":2,"y":null}] 
```

In this case, if you convert the json back to R, the missing values converted back to `NA` and the order is retained. However I found that third party languages are likely to misinterpret `null` or `"NA"`. Therefore, the default in jsonlite is to leave out missing values from the record, which, in my experience, will lead most json consumers to correctly conclude that this field is *missing*.
